### PR TITLE
add xcit model

### PR DESCRIPTION
**Pull request by learncat163**
# XCiT模型权重

[AiStudio的演示项目](https://aistudio.baidu.com/projectdetail/10611144)

[paddle权重地址](https://aistudio.baidu.com/modelsdetail/51304?modelId=51304)

## 1. 模型描述

XCiT（Cross-Covariance Image Transformers）使用 Cross-Covariance Attention（XCA）替代标准 Self-Attention，在通道维度而非 token 维度计算注意力，具备更好的可扩展性与计算效率。XCiT 还包含 Local Patch Interaction（LPI）模块增强局部信息交互，并在最后阶段使用 Class Attention（CA）聚合全局信息。

timm 已实现 XCiT 的全系列变体，但是PaddleClas还没有实现。本提交从 timm 权重转换并迁移 XCiT 到 PaddleClas，覆盖 patch size 16/8、输入分辨率 224/384、distilled 蒸馏等共 42 个变体。代码在：`ppcls/arch/backbone/model_zoo/xcit.py`。

### 2.1 整体权重概览

权重全部从 timm 权重转换而来，转换后上传至 bcebos，可通过 `pretrained=True` 自动下载。

**XCiT（timm 权重转换）**

| PaddleClas 模型 | 原始timm 模型 | paddle转换权重 | 文件大小 |
| --------------- | ------------- | -------------- | -------- |
| XCiT_nano_12_p16_224 | `timm/xcit_nano_12_p16_224.fb_in1k` | `XCiT_nano_12_p16_224.pdparams` | 11.7M |
| XCiT_nano_12_p16_224_distilled | `timm/xcit_nano_12_p16_224.fb_dist_in1k` | `XCiT_nano_12_p16_224_distilled.pdparams` | 11.7M |
| XCiT_nano_12_p16_384_distilled | `timm/xcit_nano_12_p16_384.fb_dist_in1k` | `XCiT_nano_12_p16_384_distilled.pdparams` | 11.7M |
| XCiT_tiny_12_p16_224 | `timm/xcit_tiny_12_p16_224.fb_in1k` | `XCiT_tiny_12_p16_224.pdparams` | 25.7M |
| XCiT_tiny_12_p16_224_distilled | `timm/xcit_tiny_12_p16_224.fb_dist_in1k` | `XCiT_tiny_12_p16_224_distilled.pdparams` | 25.7M |
| XCiT_tiny_12_p16_384_distilled | `timm/xcit_tiny_12_p16_384.fb_dist_in1k` | `XCiT_tiny_12_p16_384_distilled.pdparams` | 25.7M |
| XCiT_tiny_24_p16_224 | `timm/xcit_tiny_24_p16_224.fb_in1k` | `XCiT_tiny_24_p16_224.pdparams` | 46.3M |
| XCiT_tiny_24_p16_224_distilled | `timm/xcit_tiny_24_p16_224.fb_dist_in1k` | `XCiT_tiny_24_p16_224_distilled.pdparams` | 46.3M |
| XCiT_tiny_24_p16_384_distilled | `timm/xcit_tiny_24_p16_384.fb_dist_in1k` | `XCiT_tiny_24_p16_384_distilled.pdparams` | 46.3M |
| XCiT_small_12_p16_224 | `timm/xcit_small_12_p16_224.fb_in1k` | `XCiT_small_12_p16_224.pdparams` | 100.2M |
| XCiT_small_12_p16_224_distilled | `timm/xcit_small_12_p16_224.fb_dist_in1k` | `XCiT_small_12_p16_224_distilled.pdparams` | 100.2M |
| XCiT_small_12_p16_384_distilled | `timm/xcit_small_12_p16_384.fb_dist_in1k` | `XCiT_small_12_p16_384_distilled.pdparams` | 100.2M |
| XCiT_small_24_p16_224 | `timm/xcit_small_24_p16_224.fb_in1k` | `XCiT_small_24_p16_224.pdparams` | 182.0M |
| XCiT_small_24_p16_224_distilled | `timm/xcit_small_24_p16_224.fb_dist_in1k` | `XCiT_small_24_p16_224_distilled.pdparams` | 182.0M |
| XCiT_small_24_p16_384_distilled | `timm/xcit_small_24_p16_384.fb_dist_in1k` | `XCiT_small_24_p16_384_distilled.pdparams` | 182.0M |
| XCiT_medium_24_p16_224 | `timm/xcit_medium_24_p16_224.fb_in1k` | `XCiT_medium_24_p16_224.pdparams` | 322.1M |
| XCiT_medium_24_p16_224_distilled | `timm/xcit_medium_24_p16_224.fb_dist_in1k` | `XCiT_medium_24_p16_224_distilled.pdparams` | 322.1M |
| XCiT_medium_24_p16_384_distilled | `timm/xcit_medium_24_p16_384.fb_dist_in1k` | `XCiT_medium_24_p16_384_distilled.pdparams` | 322.1M |
| XCiT_large_24_p16_224 | `timm/xcit_large_24_p16_224.fb_in1k` | `XCiT_large_24_p16_224.pdparams` | 721.5M |
| XCiT_large_24_p16_224_distilled | `timm/xcit_large_24_p16_224.fb_dist_in1k` | `XCiT_large_24_p16_224_distilled.pdparams` | 721.5M |
| XCiT_large_24_p16_384_distilled | `timm/xcit_large_24_p16_384.fb_dist_in1k` | `XCiT_large_24_p16_384_distilled.pdparams` | 721.5M |
| XCiT_nano_12_p8_224 | `timm/xcit_nano_12_p8_224.fb_in1k` | `XCiT_nano_12_p8_224.pdparams` | 11.7M |
| XCiT_nano_12_p8_224_distilled | `timm/xcit_nano_12_p8_224.fb_dist_in1k` | `XCiT_nano_12_p8_224_distilled.pdparams` | 11.7M |
| XCiT_nano_12_p8_384_distilled | `timm/xcit_nano_12_p8_384.fb_dist_in1k` | `XCiT_nano_12_p8_384_distilled.pdparams` | 11.7M |
| XCiT_tiny_12_p8_224 | `timm/xcit_tiny_12_p8_224.fb_in1k` | `XCiT_tiny_12_p8_224.pdparams` | 25.6M |
| XCiT_tiny_12_p8_224_distilled | `timm/xcit_tiny_12_p8_224.fb_dist_in1k` | `XCiT_tiny_12_p8_224_distilled.pdparams` | 25.6M |
| XCiT_tiny_12_p8_384_distilled | `timm/xcit_tiny_12_p8_384.fb_dist_in1k` | `XCiT_tiny_12_p8_384_distilled.pdparams` | 25.6M |
| XCiT_tiny_24_p8_224 | `timm/xcit_tiny_24_p8_224.fb_in1k` | `XCiT_tiny_24_p8_224.pdparams` | 46.3M |
| XCiT_tiny_24_p8_224_distilled | `timm/xcit_tiny_24_p8_224.fb_dist_in1k` | `XCiT_tiny_24_p8_224_distilled.pdparams` | 46.3M |
| XCiT_tiny_24_p8_384_distilled | `timm/xcit_tiny_24_p8_384.fb_dist_in1k` | `XCiT_tiny_24_p8_384_distilled.pdparams` | 46.3M |
| XCiT_small_12_p8_224 | `timm/xcit_small_12_p8_224.fb_in1k` | `XCiT_small_12_p8_224.pdparams` | 100.1M |
| XCiT_small_12_p8_224_distilled | `timm/xcit_small_12_p8_224.fb_dist_in1k` | `XCiT_small_12_p8_224_distilled.pdparams` | 100.1M |
| XCiT_small_12_p8_384_distilled | `timm/xcit_small_12_p8_384.fb_dist_in1k` | `XCiT_small_12_p8_384_distilled.pdparams` | 100.1M |
| XCiT_small_24_p8_224 | `timm/xcit_small_24_p8_224.fb_in1k` | `XCiT_small_24_p8_224.pdparams` | 181.8M |
| XCiT_small_24_p8_224_distilled | `timm/xcit_small_24_p8_224.fb_dist_in1k` | `XCiT_small_24_p8_224_distilled.pdparams` | 181.8M |
| XCiT_small_24_p8_384_distilled | `timm/xcit_small_24_p8_384.fb_dist_in1k` | `XCiT_small_24_p8_384_distilled.pdparams` | 181.8M |
| XCiT_medium_24_p8_224 | `timm/xcit_medium_24_p8_224.fb_in1k` | `XCiT_medium_24_p8_224.pdparams` | 321.8M |
| XCiT_medium_24_p8_224_distilled | `timm/xcit_medium_24_p8_224.fb_dist_in1k` | `XCiT_medium_24_p8_224_distilled.pdparams` | 321.8M |
| XCiT_medium_24_p8_384_distilled | `timm/xcit_medium_24_p8_384.fb_dist_in1k` | `XCiT_medium_24_p8_384_distilled.pdparams` | 321.8M |
| XCiT_large_24_p8_224 | `timm/xcit_large_24_p8_224.fb_in1k` | `XCiT_large_24_p8_224.pdparams` | 720.9M |
| XCiT_large_24_p8_224_distilled | `timm/xcit_large_24_p8_224.fb_dist_in1k` | `XCiT_large_24_p8_224_distilled.pdparams` | 720.9M |
| XCiT_large_24_p8_384_distilled | `timm/xcit_large_24_p8_384.fb_dist_in1k` | `XCiT_large_24_p8_384_distilled.pdparams` | 720.9M |

### 3.1 前向推理精度对齐

通过加载转换后的 Paddle 权重，与 PyTorch timm 原版模型进行前向推理对比。测试输入为 N(0,1) 随机张量（seed=42），消除预处理干扰，纯对齐模型权重本身。目标：最大绝对误差 < 1e-4。

**Patch size 16**

| 模型 | 输入尺寸 | 最大绝对误差 | 平均绝对误差 |
| ---- | -------- | ------------ | ------------ |
| XCiT_nano_12_p16_224 | 224x224 | 2.34e-05 | 3.97e-06 |
| XCiT_nano_12_p16_224_distilled | 224x224 | 1.45e-05 | 3.03e-06 |
| XCiT_nano_12_p16_384_distilled | 384x384 | 2.44e-05 | 4.43e-06 |
| XCiT_tiny_12_p16_224 | 224x224 | 7.87e-06 | 1.48e-06 |
| XCiT_tiny_12_p16_224_distilled | 224x224 | 1.18e-05 | 2.45e-06 |
| XCiT_tiny_12_p16_384_distilled | 384x384 | 1.03e-05 | 2.45e-06 |
| XCiT_tiny_24_p16_224 | 224x224 | 5.13e-06 | 8.14e-07 |
| XCiT_tiny_24_p16_224_distilled | 224x224 | 7.15e-06 | 1.49e-06 |
| XCiT_tiny_24_p16_384_distilled | 384x384 | 1.30e-05 | 2.17e-06 |
| XCiT_small_12_p16_224 | 224x224 | 4.89e-06 | 1.15e-06 |
| XCiT_small_12_p16_224_distilled | 224x224 | 7.03e-06 | 1.60e-06 |
| XCiT_small_12_p16_384_distilled | 384x384 | 1.67e-05 | 2.31e-06 |
| XCiT_small_24_p16_224 | 224x224 | 5.33e-06 | 1.29e-06 |
| XCiT_small_24_p16_224_distilled | 224x224 | 1.80e-05 | 3.76e-06 |
| XCiT_small_24_p16_384_distilled | 384x384 | 8.11e-06 | 1.67e-06 |
| XCiT_medium_24_p16_224 | 224x224 | 6.17e-06 | 1.37e-06 |
| XCiT_medium_24_p16_224_distilled | 224x224 | 8.82e-06 | 1.48e-06 |
| XCiT_medium_24_p16_384_distilled | 384x384 | 1.62e-05 | 1.59e-06 |
| XCiT_large_24_p16_224 | 224x224 | 3.10e-06 | 7.09e-07 |
| XCiT_large_24_p16_224_distilled | 224x224 | 7.27e-06 | 1.13e-06 |
| XCiT_large_24_p16_384_distilled | 384x384 | 4.44e-06 | 9.86e-07 |

**Patch size 8**

| 模型 | 输入尺寸 | 最大绝对误差 | 平均绝对误差 |
| ---- | -------- | ------------ | ------------ |
| XCiT_nano_12_p8_224 | 224x224 | 5.07e-06 | 1.43e-06 |
| XCiT_nano_12_p8_224_distilled | 224x224 | 1.79e-05 | 3.81e-06 |
| XCiT_nano_12_p8_384_distilled | 384x384 | 1.84e-05 | 3.42e-06 |
| XCiT_tiny_12_p8_224 | 224x224 | 7.39e-06 | 1.13e-06 |
| XCiT_tiny_12_p8_224_distilled | 224x224 | 1.00e-05 | 2.11e-06 |
| XCiT_tiny_12_p8_384_distilled | 384x384 | 1.30e-05 | 2.48e-06 |
| XCiT_tiny_24_p8_224 | 224x224 | 4.29e-06 | 8.02e-07 |
| XCiT_tiny_24_p8_224_distilled | 224x224 | 6.44e-06 | 1.28e-06 |
| XCiT_tiny_24_p8_384_distilled | 384x384 | 8.17e-06 | 1.63e-06 |
| XCiT_small_12_p8_224 | 224x224 | 4.80e-06 | 1.21e-06 |
| XCiT_small_12_p8_224_distilled | 224x224 | 1.12e-05 | 1.97e-06 |
| XCiT_small_12_p8_384_distilled | 384x384 | 6.05e-06 | 1.83e-06 |
| XCiT_small_24_p8_224 | 224x224 | 5.60e-06 | 1.40e-06 |
| XCiT_small_24_p8_224_distilled | 224x224 | 9.06e-06 | 1.83e-06 |
| XCiT_small_24_p8_384_distilled | 384x384 | 8.58e-06 | 1.35e-06 |
| XCiT_medium_24_p8_224 | 224x224 | 7.60e-06 | 1.61e-06 |
| XCiT_medium_24_p8_224_distilled | 224x224 | 9.54e-06 | 1.79e-06 |
| XCiT_medium_24_p8_384_distilled | 384x384 | 1.00e-05 | 1.79e-06 |
| XCiT_large_24_p8_224 | 224x224 | 4.41e-06 | 1.09e-06 |
| XCiT_large_24_p8_224_distilled | 224x224 | 5.54e-06 | 1.33e-06 |
| XCiT_large_24_p8_384_distilled | 384x384 | 6.56e-06 | 1.37e-06 |

42/42 模型全部通过，最大绝对误差均远小于 1e-4 目标。

### 3.2 ImageNet 验证集精度验证

在 ImageNet 验证集(5万张)上验证模型的 Top-1 准确率。由于本地 ground truth 标签文件使用 ILSVRC2012 官方 1-1000 WordNet ID 顺序，与模型输出索引（0-999 按 WordNet ID 字母排序）不兼容，故采用 paddle 与 timm 在同一批图片上的预测一致率作为对比指标（严格 Top-1 匹配），目标一致率 >= 99.8%（对应 acc 误差 < 0.2%）。抽测了覆盖 patch size 16/8、分辨率 224/384、distilled 蒸馏、以及 nano/tiny/small/medium/large 各规模的 7 个典型模型：

**XCiT**

| 模型 | 输入尺寸 | 一致率 |
| ---- | -------- | ------ |
| XCiT_nano_12_p16_224 | 224x224 | 99.96% |
| XCiT_tiny_12_p16_224 | 224x224 | 99.96% |
| XCiT_tiny_12_p16_224_distilled | 224x224 | 99.98% |
| XCiT_small_24_p16_224 | 224x224 | 99.99% |
| XCiT_large_24_p16_384_distilled | 384x384 | 100.00% |
| XCiT_small_12_p8_224 | 224x224 | 99.98% |
| XCiT_medium_24_p8_384_distilled | 384x384 | 99.99% |

因为单个模型跑完5W的acc很慢，xcit的模型很大，而且数量很多。这里抽检7个典型案例。抽测的 7 个典型模型在 50000 张验证图上与 timm 预测一致率均达 99.96% 以上，满足精度要求。

### CODE REVIEW

**Comment by learncat163** [comment]
# XCiT 训练测试


## 测试配置

- **模型**: XCiT_nano_12_p16_224
- **配置文件**: `ppcls/configs/ImageNet/XCiT/XCiT_nano_12_p16_224.yaml`
- **设备**: GPU
- **训练轮数**: 50
- **批次大小**: 8
- **优化器**: AdamW (lr=2e-3, weight_decay=0.05, Cosine, LinearWarmup)
- **损失函数**: CrossEntropy (epsilon=0.1)

## 每轮训练损失

| 轮次 | CE 损失 |
|------|---------|
| 1 | 6.93491 |
| 2 | 6.62419 |
| 3 | 6.25633 |
| 4 | 5.69770 |
| 5 | 5.02006 |
| 6 | 4.40277 |
| 7 | 4.11127 |
| 8 | 4.04497 |
| 9 | 4.04698 |
| 10 | 4.00440 |
| 11 | 4.16405 |
| 12 | 4.10539 |
| 13 | 4.10414 |
| 14 | 4.08642 |
| 15 | 4.04809 |
| 16 | 4.03524 |
| 17 | 4.07761 |
| 18 | 4.05333 |
| 19 | 4.00472 |
| 20 | 4.04221 |
| 21 | 4.00285 |
| 22 | 3.98037 |
| 23 | 4.03062 |
| 24 | 3.96474 |
| 25 | 3.92820 |
| 26 | 3.96075 |
| 27 | 3.96081 |
| 28 | 4.09827 |
| 29 | 4.06664 |
| 30 | 4.04664 |
| 31 | 4.01308 |
| 32 | 4.01211 |
| 33 | 4.03664 |
| 34 | 4.01553 |
| 35 | 3.92691 |
| 36 | 3.94550 |
| 37 | 3.98205 |
| 38 | 3.93635 |
| 39 | 3.96276 |
| 40 | 3.87806 |
| 41 | 3.93809 |
| 42 | 3.95379 |
| 43 | 3.99878 |
| 44 | 3.89748 |
| 45 | 3.98369 |
| 46 | 3.99292 |
| 47 | 3.88297 |
| 48 | 3.98609 |
| 49 | 3.92344 |
| 50 | 3.96628 |

**Comment by zhangyubo0722** [comment]
解决下冲突吧，解决后合入
